src/librustc_typeck/check/writeback.rs

```diff
@@ -424,12 +420,10 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
         debug_assert_eq!(fcx_tables.local_id_root, self.tables.local_id_root);
 
         for (&def_id, c_sig) in fcx_tables.user_provided_sigs.iter() {
-            let c_sig = if let Some(c_sig) = self.tcx().lift_to_global(c_sig) {
-                c_sig
-            } else {
+            if cfg!(debug_assertions) && c_sig.has_local_value() {
```

Comment:
Hmm, can we do a crater run with all of these assertions turned on by default, to confirm they're very likely never triggered, and then remove them?

Reply:
They're effectively turned on now, and they have caught at least one bug https://github.com/rust-lang/rust/issues/46184.

I do think we should have these assertions somewhere though, to ensure that no infer types leaks out from typeck. I could just use `debug_assert` here instead of `span_bug` to make them more slim.